Malfunction tweaks/additions

atlantiscze opened the pull request:
- Added T5 manipulation tree ability - Machine Upgrade. Reduced (unlock) cost of machine overload a bit. Machine upgrade is now currently only defined for few machines ( https://docs.google.com/document/d/1Y4FYtTk1PIOhVjm2DB3ELvr3tApGAmyZCPnCuo_yiJw/edit?usp=sharing ).
- Removed Turrets focus enhancer hardware. Functionality shifted to machine upgrade.
- Added new hardware - "Quantum Knowledge Databank". It is a single use hardware that advances research in all trees by one level. Best used for endgame.
- Nuke station timer increased from 2 minutes to 5 minutes, to give the crew at least some time to react to the AI.
- Added new research tree - passive. Abilities are as below:
- T1: Intellicard interception - Allows the AI to prevent carding. Toggleable. Malf AI can now be carded by default, unless this ability is researched and enabled.
- T2: Subtle Algorithms - Stealth ability, removes BSOD and access lockout effects from hacked APCs, at the cost of reducing CPU generation by 50%. Toggleable.
- T3: Relay Suppression - Prevents advanced/elite encryption hack from failing or critically failing. Prevents System Override (nuke hack) from broadcasting centcom warning messages. Automatically enabled on-unlock.
- T4: Relay Override - Intercepts off-station faxes and emergency transmissions preventing them from reaching central/solgov/whatever (they are marked "Intercepted by [ai name]", but staff still see them. Shouldn't respond to them though.). Automatically enabled on-unlock.

atlantiscze committed:
Malfunction tweaks/additions
- Added T5 manipulation tree ability - Machine Upgrade. Reduced cost of machine overload a bit. Machine upgrade is now currently only defined for few machines ( https://docs.google.com/document/d/1Y4FYtTk1PIOhVjm2DB3ELvr3tApGAmyZCPnCuo_yiJw/edit?usp=sharing ).
- Removed Turrets focus enhancer hardware. Functionality shifted to machine upgrade.
- Added new hardware - "Quantum Knowledge Databank". It is a single use hardware that advances research in all trees by one level. Best used for endgame.
- Nuke station timer increased from 2 minutes to 5 minutes, to give the crew at least some time to react to the AI.
- Added new research tree - passive. Abilities are as below:
- T1: Intellicard interception - Allows the AI to prevent carding. Toggleable. Malf AI can now be carded by default, unless this ability is researched and enabled.
- T2: Subtle Algorithms - Stealth ability, removes BSOD and access lockout effects from hacked APCs, at the cost of reducing CPU generation by 50%. Toggleable.
- T3: Relay Suppression - Prevents advanced/elite encryption hack from failing or critically failing. Prevents System Override (nuke hack) from broadcasting centcom warning messages. Automatically enabled on-unlock.
- T4: Relay Override - Intercepts off-station faxes and emergency transmissions preventing them from reaching central/solgov/whatever (they are marked "Intercepted by [ai name]", but staff still see them. Shouldn't respond to them though.). Automatically enabled on-unlock.

Kelenius commented on code/game/gamemodes/malfunction/newmalf_ability_trees/HARDWARE.dm:
```diff
@@ -62,6 +62,34 @@
+
+	var/datum/malf_hardware/instant_research/HW = user.hardware
+	if(HW.spent)
+		to_chat(user, "You attempt to activate your hardware piece, but it does not work. It must be damaged.")
```
Don't leave a vague message, tell the player directly that it was already used.

atlantiscze replied:
That can be done.

Kelenius commented on code/game/objects/items/devices/aicard.dm:
```diff
@@ -97,7 +97,7 @@
 		return 0
 
-	if(ai.malfunctioning)
+	if(ai.malfunctioning && ai.uncardable)
```
Why not change it to just ai.uncardable?

atlantiscze replied:
Can do that.

Kelenius commented on code/modules/power/smes_construction.dm:
```diff
@@ -154,6 +165,10 @@
 			input_level_max += C.IOCapacity
 			output_level_max += C.IOCapacity
+		if(malf_upgraded)
+			capacity *= 1.2
```
Since its primary use is probably going to be AI's own SMES, capacity is a lot more important than input and output. 1.2 is nothing.

atlantiscze replied:
Original idea here was actually to allow for higher damage when electrifying doors, or for more power for upgraded turrets (which become a decent power hog when you have bunch of them firing at once).

Kelenius commented on code/game/gamemodes/malfunction/newmalf_ability_trees/tree_manipulation.dm:
```diff
@@ -2,38 +2,45 @@
+
+/datum/malf_research_ability/manipulation/machine_upgrade
+	ability = new/datum/game_mode/malfunction/verb/machine_upgrade()
 	price = 7500
```
50 + 1200 + 3000 + 6000 + 7500 = 17750. The AI gains 0.1 + 0.04 * APC amount per second. Let's say we have 50 hacked APCs. That's 2.1 per second. ~8500 seconds or ~140 minutes. More than two hours on one tree, that's assuming you already have 50 APCs and don't use your CPU power on anything else, like hacking APCs. Even with the hardware CPU upgrade this is still a lot of time to get... x-ray cameras and EMP-proof APCs? Really?

atlantiscze replied:
The price feels apropriate to me. It is the same upgrade that lets you double a SMES's transfer rate (effectively increasing electrification damage considerably), or upgrade turrets to have double rate of fire, higher damage and limited self repair.

Yes, it doesn't work on all machines right now, but it is doable. It is, in fact, a matter of implementing a single proc on any obj/machinery/ subtype, similarly to how emagging works. Imagine it to be a re-usable emag for malf AIs.

Furthermore, don't forget that research isn't only about getting the T5 ability. T4 Machine Overload is also extremely useful, as are the other lower tier abilities.

The whole intent of research system is that you shouldn't be able (or it should be very hard) to research everything during your average round. T4/T5 abilities (except for system override which is relatively cheap) are best researched while performing system override, as the growing amount of APCs will provide a steady amount of CPU time for you.

Kelenius replied:
>It is the same upgrade that lets you double a SMES's transfer rate (effectively increasing electrification damage considerably)

That's not how it works. Besides, after the first electrified doors, people quickly yoink gloves or stop using doors altogether.

>or upgrade turrets to have double rate of fire, higher damage and limited self repair

And they still can be easily kited. Useless. Also, turrets currently don't shoot though glass - https://github.com/PolarisSS13/Polaris/issues/1089

>The whole intent of research system is that you shouldn't be able (or it should be very hard) to research everything during your average round

1. Why 2. At this rate, you won't even be able to research this branch entirely at least until the very end of the round

>T4/T5 abilities (except for system override which is relatively cheap) are best researched while performing system override

...if you are doing the override, round is most likely about to end.

atlantiscze replied:
Electrify: That *IS* how it works. See powernet code - burn damage is calculated from amount of energy. While it's not 1:1 rate, more energy = more damage in general.

Turrets: Ability to kite turrets is not a fault of this PR. Neither is shooting through glass.

Research, Why: To introduce and element of choice and specialisation.

Research, Time: This depends on your hardware. The quantum databank for example lets you, theoretically, unlock about 15-20 000 CPU units worth of abilities (it advances all abilities by one tier), alternatively, dual CPU boosts general CPU generation.

Override: Override isn't the end. In fact, one of the best parts of malf rounds i've had was after override, where the AI can use ability to toggle the nuke as a negotiating method against the crew.

mwerezak replied:
Personally I have always felt that the Malf AI research costs were grossly overpriced.

Not being able to research absolutely everything is fine, so there is some choice involved in what to research. I think what you don't get is the timing. If you research something by the time the round is ending, it's mostly useless. So what you should be thinking is about is not being able to research absolutely everything by mid-round or so. 

I would aim to make it so that one tree takes about an hour, so the AI player has at least one tree to work with by the time things start going. Two trees two hours in, already late in the round. Three trees by the end of the round, at which point it doesn't matter anymore.

It's very important to remember that you cannot research while also hacking APCs, this fact has huge balance implications. So you also need to consider the time it takes to acquire enough APCs to have a decent research rate.

atlantiscze replied:
@mwerezak I can reduce the research costs a bit overall as an experiment, but i'd also like to point out that you can research while hacking APCs, you can't use your other abilities while using basic hack, however. Once you have more than ~5 APCs you'll start generating CPU time faster than you use it when only hacking APCs. Anything excess will go into research.

Kelenius replied:
You could also remove tiers. Because right now "you can't get everything you want" can easily lead to "you can't get two specific abilities even if you don't want anything else because they are at the end of the trees".

Kelenius commented on code/game/gamemodes/malfunction/newmalf_ability_trees/tree_passive.dm:
```diff
@@ -0,0 +1,81 @@
+
+/datum/malf_research_ability/passive/relay_override
+	ability = null
+	price = 5000
```
This is also wayyyyyy too expensive, for an ability that merely stops admin intervention.

atlantiscze replied:
I admit, for this particular one it is true, which is because the ability was originally actually intended to allow for actual interception: the AI would have seen the messages, and would have ability to reply to them. Unfortunately, i've had to abandon that as replies (especially fax replies and viewing/editing) seem to be quite tightly tied to admin stuff.

I'm not saying it's not doable, i just didn't have time to do that. I'm most likely going to drop this price a little.

atlantiscze committed:
Normalizes (and overall mostly lowers) ability research costs
- T1: 250 CPU (minor increase), T2 1000 CPU (minor decrease), T3 2000 CPU (decent decrease), T4 and T5 4000 CPU (also decent decrease)
- Exception: Basic Encryption Hack, as you don't have income from APCs without this, i've left it's research price as only 25.
- For most trees this means they're cheaper now. The only tree that's overall more expensive than it used to be is networking, which was much cheaper than the other ones before. Overall, if you count together all research prices of all trees the value will be lower by ~20-30%.
- Tweaked electrical pulse ability a bit. APC break chance was increased from 0.01% per APC to 2.5% per APC. Added check so it shouldn't affect the AI's APC.

comma committed:
Merge branch 'dev' into 2017-02-08-SystemOverrideCompleted

Minijar commented:
Bump?

atlantiscze committed:
Merge branch 'dev' into 2017-02-08-SystemOverrideCompleted

Techhead0 commented on code/game/gamemodes/malfunction/newmalf_ability_trees/tree_networking.dm:
```diff
@@ -199,15 +198,17 @@
-	command_announcement.Announce("Our system administrators just reported that we've been locked out from your control network. Whoever did this now has full access to [using_map.station_name]'s systems.", "Network Administration Center")
+  
+	if(user.hack_can_fail)
+	  command_announcement.Announce("Our system administrators just reported that we've been locked out from your control network. Whoever did this now has full access to [using_map.station_name]'s systems.", "Network Administration Center")
```
Noticed you using spaces for code indentation.

comma commented:
Abandoned
